File: frontend/src/services/university.ts
import { supabase } from '../lib/supabase'
import type { UniSubject, UniSchedule, UniExam, UniSemester } from '../types/uni.types'

async function uid() {
  const { data: { session } } = await supabase.auth.getSession()
  if (!session?.user) throw new Error('Não autenticado')
  return session.user.id
}

// ── Subjects ─────────────────────────────────────────────────────────────────

function mapSubject(r: any): UniSubject {
  return {
    id: r.id, name: r.name, professor: r.professor, credits: r.credits,
    grade: r.grade ?? null, attendance: r.attendance,
    absences: r.absences ?? 0, maxAbsences: r.max_absences ?? 0,
    subjectStatus: r.subject_status ?? 'open',
    startDate: r.start_date ?? null, endDate: r.end_date ?? null,
    colorIdx: r.color_idx, icon: r.icon, semester: r.semester, created_at: r.created_at,
  }
}

export async function fetchSubjects(): Promise<UniSubject[]> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_subjects').select('*').eq('user_id', userId).order('created_at')
  if (error) throw new Error(error.message)
  return (data ?? []).map(mapSubject)
}

export async function createSubject(p: Omit<UniSubject, 'id' | 'created_at'>): Promise<UniSubject> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_subjects')
    .insert({ user_id: userId, name: p.name.trim(), professor: p.professor.trim(), credits: p.credits, grade: p.grade, attendance: p.attendance, absences: p.absences, max_absences: p.maxAbsences, subject_status: p.subjectStatus, start_date: p.startDate || null, end_date: p.endDate || null, color_idx: p.colorIdx, icon: p.icon, semester: p.semester.trim() })
    .select().single()
  if (error) throw new Error(error.message)
  return mapSubject(data)
}

export async function updateSubject(id: string, p: Partial<Omit<UniSubject, 'id' | 'created_at'>>): Promise<UniSubject> {
  const u: Record<string, unknown> = {}
  if (p.name       !== undefined) u.name       = p.name.trim()
  if (p.professor  !== undefined) u.professor  = p.professor.trim()
  if (p.credits    !== undefined) u.credits    = p.credits
  if (p.grade      !== undefined) u.grade      = p.grade
  if (p.attendance    !== undefined) u.attendance    = p.attendance
  if (p.absences      !== undefined) u.absences      = p.absences
  if (p.maxAbsences   !== undefined) u.max_absences  = p.maxAbsences
  if (p.subjectStatus !== undefined) u.subject_status = p.subjectStatus
  if (p.startDate  !== undefined) u.start_date = p.startDate || null
  if (p.endDate    !== undefined) u.end_date   = p.endDate || null
  if (p.colorIdx   !== undefined) u.color_idx  = p.colorIdx
  if (p.icon       !== undefined) u.icon       = p.icon
  if (p.semester   !== undefined) u.semester   = p.semester.trim()
  const { data, error } = await supabase.from('uni_subjects').update(u).eq('id', id).select().single()
  if (error) throw new Error(error.message)
  return mapSubject(data)
}

export async function deleteSubject(id: string): Promise<void> {
  const { error } = await supabase.from('uni_subjects').delete().eq('id', id)
  if (error) throw new Error(error.message)
}

// ── Schedule ──────────────────────────────────────────────────────────────────

function mapSchedule(r: any): UniSchedule {
  return {
    id: r.id, subjectName: r.subject_name, room: r.room,
    dayOfWeek: r.day_of_week, timeStart: r.time_start,
    colorIdx: r.color_idx, semester: r.semester, created_at: r.created_at,
  }
}

export async function fetchSchedule(): Promise<UniSchedule[]> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_schedule').select('*').eq('user_id', userId)
  if (error) throw new Error(error.message)
  return (data ?? []).map(mapSchedule)
}

export async function createScheduleItem(p: Omit<UniSchedule, 'id' | 'created_at'>): Promise<UniSchedule> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_schedule')
    .insert({ user_id: userId, subject_name: p.subjectName.trim(), room: p.room.trim(), day_of_week: p.dayOfWeek, time_start: p.timeStart, color_idx: p.colorIdx, semester: p.semester.trim() })
    .select().single()
  if (error) throw new Error(error.message)
  return mapSchedule(data)
}

export async function deleteScheduleItem(id: string): Promise<void> {
  const { error } = await supabase.from('uni_schedule').delete().eq('id', id)
  if (error) throw new Error(error.message)
}

// ── Exams ─────────────────────────────────────────────────────────────────────

function mapExam(r: any): UniExam {
  return {
    id: r.id, subject: r.subject, examDate: r.exam_date,
    type: r.type, description: r.description, status: r.status, created_at: r.created_at,
  }
}

export async function fetchExams(): Promise<UniExam[]> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_exams').select('*').eq('user_id', userId).order('exam_date')
  if (error) throw new Error(error.message)
  return (data ?? []).map(mapExam)
}

export async function createExam(p: Omit<UniExam, 'id' | 'created_at'>): Promise<UniExam> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_exams')
    .insert({ user_id: userId, subject: p.subject.trim(), exam_date: p.examDate, type: p.type.trim(), description: p.description.trim(), status: p.status })
    .select().single()
  if (error) throw new Error(error.message)
  return mapExam(data)
}

export async function updateExam(id: string, p: Partial<Omit<UniExam, 'id' | 'created_at'>>): Promise<UniExam> {
  const u: Record<string, unknown> = {}
  if (p.subject     !== undefined) u.subject     = p.subject.trim()
  if (p.examDate    !== undefined) u.exam_date   = p.examDate
  if (p.type        !== undefined) u.type        = p.type.trim()
  if (p.description !== undefined) u.description = p.description.trim()
  if (p.status      !== undefined) u.status      = p.status
  const { data, error } = await supabase.from('uni_exams').update(u).eq('id', id).select().single()
  if (error) throw new Error(error.message)
  return mapExam(data)
}

export async function toggleExamStatus(id: string, status: 'pending' | 'done'): Promise<UniExam> {
  const { data, error } = await supabase.from('uni_exams').update({ status }).eq('id', id).select().single()
  if (error) throw new Error(error.message)
  return mapExam(data)
}

export async function deleteExam(id: string): Promise<void> {
  const { error } = await supabase.from('uni_exams').delete().eq('id', id)
  if (error) throw new Error(error.message)
}

// ── Semesters ─────────────────────────────────────────────────────────────────

function mapSemester(r: any): UniSemester {
  return {
    id: r.id, name: r.name,
    startDate: r.start_date ?? null, endDate: r.end_date ?? null,
    isCurrent: r.is_current ?? false, created_at: r.created_at,
  }
}

export async function fetchSemesters(): Promise<UniSemester[]> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_semesters').select('*').eq('user_id', userId).order('created_at', { ascending: false })
  if (error) throw new Error(error.message)
  return (data ?? []).map(mapSemester)
}

export async function createSemester(p: { name: string; startDate: string | null; endDate: string | null }): Promise<UniSemester> {
  const userId = await uid()
  const { data, error } = await supabase.from('uni_semesters')
    .insert({ user_id: userId, name: p.name.trim(), start_date: p.startDate || null, end_date: p.endDate || null, is_current: false })
    .select().single()
  if (error) throw new Error(error.message)
  return mapSemester(data)
}

export async function setCurrentSemester(id: string, userId: string): Promise<void> {
  await supabase.from('uni_semesters').update({ is_current: false }).eq('user_id', userId)
  const { error } = await supabase.from('uni_semesters').update({ is_current: true }).eq('id', id)
  if (error) throw new Error(error.message)
}

export async function encerrarSemesterService(id: string): Promise<void> {
  const { error } = await supabase.from('uni_semesters').update({ is_current: false }).eq('id', id)
  if (error) throw new Error(error.message)
}

export async function deleteSemester(id: string): Promise<void> {
  const { error } = await supabase.from('uni_semesters').delete().eq('id', id)
  if (error) throw new Error(error.message)
}

File: backend/handlers/university.go
package handlers

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/gmourazz/bee-planner/config"
	"github.com/gmourazz/bee-planner/middleware"
)

// ── Matérias ──────────────────────────────────────────────────────────────────

// GET /api/university/subjects
func ListSubjects(c *gin.Context) {
	user := middleware.GetUser(c)

	body, status, err := config.Supabase.From("uni_subjects").
		Select("*").Eq("user_id", user.ID).Order("created_at", true).Get()
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao buscar matérias"})
		return
	}

	var rows []map[string]any
	json.Unmarshal(body, &rows)
	if rows == nil {
		rows = []map[string]any{}
	}
	c.JSON(http.StatusOK, rows)
}

// POST /api/university/subjects
func CreateSubject(c *gin.Context) {
	user := middleware.GetUser(c)

	var body struct {
		Name          string   `json:"name"`
		Professor     string   `json:"professor"`
		Credits       *int     `json:"credits"`
		Grade         *float64 `json:"grade"`
		Attendance    *float64 `json:"attendance"`
		Absences      *int     `json:"absences"`
		MaxAbsences   *int     `json:"max_absences"`
		SubjectStatus string   `json:"subject_status"`
		StartDate     string   `json:"start_date"`
		EndDate       string   `json:"end_date"`
		ColorIdx      *int     `json:"color_idx"`
		Icon          string   `json:"icon"`
		Semester      string   `json:"semester"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if strings.TrimSpace(body.Name) == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Nome obrigatório"})
		return
	}

	status_ := body.SubjectStatus
	if status_ == "" {
		status_ = "open"
	}
	absences := 0
	if body.Absences != nil {
		absences = *body.Absences
	}
	maxAbsences := 0
	if body.MaxAbsences != nil {
		maxAbsences = *body.MaxAbsences
	}

	resp, status, err := config.Supabase.From("uni_subjects").Insert(map[string]any{
		"user_id":        user.ID,
		"name":           strings.TrimSpace(body.Name),
		"professor":      body.Professor,
		"credits":        body.Credits,
		"grade":          body.Grade,
		"attendance":     body.Attendance,
		"absences":       absences,
		"max_absences":   maxAbsences,
		"subject_status": status_,
		"start_date":     nilIfEmpty(body.StartDate),
		"end_date":       nilIfEmpty(body.EndDate),
		"color_idx":      body.ColorIdx,
		"icon":           body.Icon,
		"semester":       body.Semester,
	})
	if err != nil || status >= 400 {
		var errResp map[string]any
		json.Unmarshal(resp, &errResp)
		c.JSON(http.StatusInternalServerError, errResp)
		return
	}

	var rows []map[string]any
	json.Unmarshal(resp, &rows)
	if len(rows) == 0 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao criar matéria"})
		return
	}
	c.JSON(http.StatusCreated, rows[0])
}

// PUT /api/university/subjects/:id
func UpdateSubject(c *gin.Context) {
	user := middleware.GetUser(c)
	id := c.Param("id")

	var body struct {
		Name          *string  `json:"name"`
		Professor     *string  `json:"professor"`
		Credits       *int     `json:"credits"`
		Grade         *float64 `json:"grade"`
		Attendance    *float64 `json:"attendance"`
		Absences      *int     `json:"absences"`
		MaxAbsences   *int     `json:"max_absences"`
		SubjectStatus *string  `json:"subject_status"`
		StartDate     *string  `json:"start_date"`
		EndDate       *string  `json:"end_date"`
		ColorIdx      *int     `json:"color_idx"`
		Icon          *string  `json:"icon"`
		Semester      *string  `json:"semester"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	updates := map[string]any{}
	if body.Name != nil          { updates["name"]           = strings.TrimSpace(*body.Name) }
	if body.Professor != nil     { updates["professor"]      = *body.Professor }
	if body.Credits != nil       { updates["credits"]        = *body.Credits }
	if body.Grade != nil         { updates["grade"]          = *body.Grade }
	if body.Attendance != nil    { updates["attendance"]     = *body.Attendance }
	if body.Absences != nil      { updates["absences"]       = *body.Absences }
	if body.MaxAbsences != nil   { updates["max_absences"]   = *body.MaxAbsences }
	if body.SubjectStatus != nil { updates["subject_status"] = *body.SubjectStatus }
	if body.StartDate != nil     { updates["start_date"]     = nilIfEmpty(*body.StartDate) }
	if body.EndDate != nil       { updates["end_date"]       = nilIfEmpty(*body.EndDate) }
	if body.ColorIdx != nil      { updates["color_idx"]      = *body.ColorIdx }
	if body.Icon != nil          { updates["icon"]           = *body.Icon }
	if body.Semester != nil      { updates["semester"]       = *body.Semester }

	resp, status, err := config.Supabase.From("uni_subjects").
		Eq("id", id).Eq("user_id", user.ID).Update(updates)
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao atualizar matéria"})
		return
	}

	var rows []map[string]any
	json.Unmarshal(resp, &rows)
	if len(rows) == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Matéria não encontrada"})
		return
	}
	c.JSON(http.StatusOK, rows[0])
}

// DELETE /api/university/subjects/:id
func DeleteSubject(c *gin.Context) {
	user := middleware.GetUser(c)
	id := c.Param("id")

	_, status, err := config.Supabase.From("uni_subjects").
		Eq("id", id).Eq("user_id", user.ID).Delete()
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao deletar matéria"})
		return
	}
	c.Status(http.StatusNoContent)
}

// ── Grade horária ─────────────────────────────────────────────────────────────

// GET /api/university/schedule
func ListSchedule(c *gin.Context) {
	user := middleware.GetUser(c)

	body, status, err := config.Supabase.From("uni_schedule").
		Select("*").Eq("user_id", user.ID).Get()
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao buscar grade"})
		return
	}

	var rows []map[string]any
	json.Unmarshal(body, &rows)
	if rows == nil {
		rows = []map[string]any{}
	}
	c.JSON(http.StatusOK, rows)
}

// POST /api/university/schedule
func CreateClass(c *gin.Context) {
	user := middleware.GetUser(c)

	var body struct {
		SubjectName string `json:"subject_name"`
		Room        string `json:"room"`
		DayOfWeek   *int   `json:"day_of_week"`
		TimeStart   string `json:"time_start"`
		ColorIdx    *int   `json:"color_idx"`
		Semester    string `json:"semester"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if strings.TrimSpace(body.SubjectName) == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Nome da matéria obrigatório"})
		return
	}

	resp, status, err := config.Supabase.From("uni_schedule").Insert(map[string]any{
		"user_id":      user.ID,
		"subject_name": strings.TrimSpace(body.SubjectName),
		"room":         body.Room,
		"day_of_week":  body.DayOfWeek,
		"time_start":   body.TimeStart,
		"color_idx":    body.ColorIdx,
		"semester":     body.Semester,
	})
	if err != nil || status >= 400 {
		var errResp map[string]any
		json.Unmarshal(resp, &errResp)
		c.JSON(http.StatusInternalServerError, errResp)
		return
	}

	var rows []map[string]any
	json.Unmarshal(resp, &rows)
	if len(rows) == 0 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao criar aula"})
		return
	}
	c.JSON(http.StatusCreated, rows[0])
}

// DELETE /api/university/schedule/:id
func DeleteClass(c *gin.Context) {
	user := middleware.GetUser(c)
	id := c.Param("id")

	_, status, err := config.Supabase.From("uni_schedule").
		Eq("id", id).Eq("user_id", user.ID).Delete()
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao deletar aula"})
		return
	}
	c.Status(http.StatusNoContent)
}

// ── Provas ────────────────────────────────────────────────────────────────────

// GET /api/university/exams
func ListExams(c *gin.Context) {
	user := middleware.GetUser(c)

	body, status, err := config.Supabase.From("uni_exams").
		Select("*").Eq("user_id", user.ID).Order("exam_date", true).Get()
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao buscar provas"})
		return
	}

	var rows []map[string]any
	json.Unmarshal(body, &rows)
	if rows == nil {
		rows = []map[string]any{}
	}
	c.JSON(http.StatusOK, rows)
}

// POST /api/university/exams
func CreateExam(c *gin.Context) {
	user := middleware.GetUser(c)

	var body struct {
		Subject     string `json:"subject"`
		ExamDate    string `json:"exam_date"`
		Type        string `json:"type"`
		Description string `json:"description"`
		Status      string `json:"status"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if strings.TrimSpace(body.Subject) == "" || body.ExamDate == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Matéria e data obrigatórias"})
		return
	}

	examStatus := body.Status
	if examStatus == "" {
		examStatus = "pending"
	}

	resp, status, err := config.Supabase.From("uni_exams").Insert(map[string]any{
		"user_id":     user.ID,
		"subject":     strings.TrimSpace(body.Subject),
		"exam_date":   body.ExamDate,
		"type":        body.Type,
		"description": body.Description,
		"status":      examStatus,
	})
	if err != nil || status >= 400 {
		var errResp map[string]any
		json.Unmarshal(resp, &errResp)
		c.JSON(http.StatusInternalServerError, errResp)
		return
	}

	var rows []map[string]any
	json.Unmarshal(resp, &rows)
	if len(rows) == 0 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao criar prova"})
		return
	}
	c.JSON(http.StatusCreated, rows[0])
}

// PATCH /api/university/exams/:id/toggle
func ToggleExam(c *gin.Context) {
	user := middleware.GetUser(c)
	id := c.Param("id")

	var body struct {
		Status string `json:"status"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	if body.Status != "pending" && body.Status != "done" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Status inválido"})
		return
	}

	resp, status, err := config.Supabase.From("uni_exams").
		Eq("id", id).Eq("user_id", user.ID).Update(map[string]any{"status": body.Status})
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao atualizar prova"})
		return
	}

	var rows []map[string]any
	json.Unmarshal(resp, &rows)
	if len(rows) == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Prova não encontrada"})
		return
	}
	c.JSON(http.StatusOK, rows[0])
}

// DELETE /api/university/exams/:id
func DeleteExam(c *gin.Context) {
	user := middleware.GetUser(c)
	id := c.Param("id")

	_, status, err := config.Supabase.From("uni_exams").
		Eq("id", id).Eq("user_id", user.ID).Delete()
	if err != nil || status >= 400 {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao deletar prova"})
		return
	}
	c.Status(http.StatusNoContent)
}
